src/main/webapp/app/entities/result/updating-result.component.ts

```diff
@@ -19,110 +17,59 @@ import { ProgrammingExercise } from 'app/entities/programming-exercise';
-        private accountService: AccountService,
-        private translate: TranslateService,
-        private http: HttpClient,
-        private participationWebsocketService: ParticipationWebsocketService,
-    ) {}
-
-    ngOnInit(): void {
-        if (!this.participation || !this.participation.id) {
-            return;
-        }
+    constructor(private accountService: AccountService, private participationWebsocketService: ParticipationWebsocketService) {}
 
-        if (this.result) {
-            const exercise = this.participation.exercise;
-            if (exercise && exercise.type === ExerciseType.PROGRAMMING) {
-                this.subscribeForNewResults(exercise as ProgrammingExercise);
+    ngOnChanges(changes: SimpleChanges) {
+        if (hasParticipationChanged(changes)) {
+            // Sort participation results by completionDate desc.
+            if (this.participation.results) {
+                this.participation.results = _orderBy(this.participation.results, 'completionDate', 'desc');
             }
-        } else {
-            const exercise = this.participation.exercise;
-
-            if (this.participation.results && this.participation.results.length > 0) {
-                if (exercise && exercise.type === ExerciseType.MODELING) {
-                    // sort results by completionDate descending to ensure the newest result is shown
-                    // this is important for modeling exercises since students can have multiple tries
-                    // think about if this should be used for all types of exercises
-                    this.participation.results.sort((r1: Result, r2: Result) => {
-                        if (r1.completionDate > r2.completionDate) {
-                            return -1;
-                        }
-                        if (r1.completionDate < r2.completionDate) {
-                            return 1;
-                        }
-                        return 0;
-                    });
-                }
-                // Make sure result and participation are connected
-                this.result = this.participation.results[0];
-                this.result.participation = this.participation;
+            // The latest result is the first rated result in the sorted array (=newest) or any result if the option is active to show ungraded results.
+            const latestResult = this.participation.results && this.participation.results.find(({ rated }) => this.showUngradedResults || rated === true);
+            if (latestResult) {
+                // Make sure that the participation result is connected to the newest result.
+                this.result = { ...latestResult, participation: this.participation };
             }
 
-            this.subscribeForNewResults(exercise);
+            this.subscribeForNewResults();
         }
     }
 
-    subscribeForNewResults(exercise: Exercise) {
-        this.accountService.identity().then(user => {
```

Comment:
I don't think there is a reason to have a permission check here, because the websocket subscription is based on the participation and user. Or is there a reason?

Reply:
Also this is the only place in the whole app that I'm aware of that does this.

Reply:
I noticed that in a lot of places on the server, the result is not addressed to the user (convertAndSendToUser), but just send to the participation (convertAndSend).
This is not a concrete issue at the moment as the participations have authorization checks of their own, but not ideal.

Reply:
The reason is that we also want to inform instructors about this. Normal users typically only listen to their own participation so this should not be a problem

Reply:
You are right!